
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Sequence Models &#8212; BEBi 205: Deep Learning for Biological Data</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="canonical" href="vanvalenlab.github.io/bebi205/notebooks/sequence-model.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Cell Segmentation" href="segmentation.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">BEBi 205: Deep Learning for Biological Data</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../syllabus.html">
   Syllabus
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../schedule.html">
   Course Schedule
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../final-project.html">
   Class Project
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cluster-access.html">
   Caltech Cluster Access
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../notebooks-toc.html">
   Course Notebooks
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="data-science-in-python-key.html">
     Data Science in Python
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="images.html">
     Images as Data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="sequences-key.html">
     Sequences
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="linear-classifier-key.html">
     The Linear Classifier
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="tf-dataset-key.html">
     TensorFlow Data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="tf-classifier.html">
     Image classification with TensorFlow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="optimizer.html">
     Optimizers
    </a>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../vision-model-toc.html">
     Vision Models
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
    <label for="toctree-checkbox-2">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="spot-detection.html">
       Spot Detection
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="segmentation.html">
       Cell Segmentation
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Sequence Models
    </a>
   </li>
  </ul>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/notebooks/sequence-model.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/vanvalenlab/bebi205"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/vanvalenlab/bebi205/issues/new?title=Issue%20on%20page%20%2Fnotebooks/sequence-model.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/vanvalenlab/bebi205/edit/master/bebi205/notebooks/sequence-model.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/vanvalenlab/bebi205/master?urlpath=tree/bebi205/notebooks/sequence-model.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-preparation">
   Data Preparation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#process-metadata">
   Process Metadata
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#process-sequences">
   Process Sequences
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#build-model">
   Build Model
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#sample-embeddings">
   Sample Embeddings
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="sequence-models">
<h1>Sequence Models<a class="headerlink" href="#sequence-models" title="Permalink to this headline">¶</a></h1>
<p>Last updated Apr 27, 2021</p>
<p><a class="reference external" href="https://colab.research.google.com/github/vanvalenlab/bebi205/blob/master/bebi205/notebooks/sequence-model.ipynb"><img alt="Open in Colab" src="https://img.shields.io/static/v1?logo=google-colab&amp;message=Open%20in%20colab&amp;color=blue&amp;label=%20&amp;labelColor=5c5c5c" /></a></p>
<p>We are going to reproduce the models published in Hie et al. 2021. The fasta file that we will initially work with is hosted in the class bucket, but the full dataset can be donwloaded using the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget http://cb.csail.mit.edu/cb/viral-mutation/data.tar.gz
tar xvf data.tar.gz
</pre></div>
</div>
<p>Hie, B., Zhong, E. D., Berger, B., &amp; Bryson, B. (2021). <a class="reference external" href="https://doi.org/10.1126/science.abd7331">Learning the language of viral evolution and escape.</a> <em>Science</em>, 371(6526), 284-288.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip install biopython
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">tensorflow</span> <span class="k">as</span> <span class="nn">tf</span>
<span class="kn">import</span> <span class="nn">umap</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="data-preparation">
<h2>Data Preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">¶</a></h2>
<div class="cell tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>wget https://storage.googleapis.com/datasets-spring2021/ird_influenzaA_HA_allspecies.fa
</pre></div>
</div>
</div>
</div>
<p>First we’ll collect the sequences and their descriptions in two separate lists.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># File paths for influenze sequences and metadata</span>
<span class="n">fpath</span> <span class="o">=</span> <span class="s1">&#39;ird_influenzaA_HA_allspecies.fa&#39;</span> 

<span class="n">seqs</span><span class="p">,</span> <span class="n">dscr</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="p">[],[],[]</span>
<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fpath</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">):</span>
    <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
    <span class="n">dscr</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
    <span class="n">length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">))</span>

<span class="nb">len</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>94560
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="process-metadata">
<h2>Process Metadata<a class="headerlink" href="#process-metadata" title="Permalink to this headline">¶</a></h2>
<p>Each sequence carries a text based description that we collected in a list. We can pass it into a pandas dataframe and then parse out useful information.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;description&#39;</span><span class="p">:</span><span class="n">dscr</span><span class="p">})</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>gb:K00429|ncbiId:AAR96248.1|UniProtKB:Q6LEJ4|O...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>gb:J02176|ncbiId:AAA43209.1|UniProtKB:P03454|O...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>gb:CY021709|ncbiId:ABP49327.1|UniProtKB:A4U6V2...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>gb:CY020285|ncbiId:ABO38054.1|UniProtKB:A4GBW6...</td>
    </tr>
    <tr>
      <th>4</th>
      <td>gb:CY083910|ncbiId:ADX98960.1|UniProtKB:F0TT51...</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([&#39;gb:K00429|ncbiId:AAR96248.1|UniProtKB:Q6LEJ4|Organism:Influenza A virus (A/seal/Mass/1/1980(H7N7))|Strain Name:A/seal/Mass/1/1980|Protein Name:HA Hemagglutinin|Gene Symbol:HA|Segment:4|Subtype:H7N7|Host:Sea Mammal&#39;],
      dtype=object)
</pre></div>
</div>
</div>
</div>
<p>Pandas has useful <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/series.html#string-handling">functions for parsing string data</a> which we will leverate to turn the description into a table of metadata.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract column names</span>
<span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)]</span>
<span class="n">cols</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;gb&#39;,
 &#39;ncbiId&#39;,
 &#39;UniProtKB&#39;,
 &#39;Organism&#39;,
 &#39;Strain Name&#39;,
 &#39;Protein Name&#39;,
 &#39;Gene Symbol&#39;,
 &#39;Segment&#39;,
 &#39;Subtype&#39;,
 &#39;Host&#39;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Split description at | and expand into separate columns</span>
<span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;description&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">,</span><span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>description</th>
      <th>gb</th>
      <th>ncbiId</th>
      <th>UniProtKB</th>
      <th>Organism</th>
      <th>Strain Name</th>
      <th>Protein Name</th>
      <th>Gene Symbol</th>
      <th>Segment</th>
      <th>Subtype</th>
      <th>Host</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>gb:K00429|ncbiId:AAR96248.1|UniProtKB:Q6LEJ4|O...</td>
      <td>gb:K00429</td>
      <td>ncbiId:AAR96248.1</td>
      <td>UniProtKB:Q6LEJ4</td>
      <td>Organism:Influenza A virus (A/seal/Mass/1/1980...</td>
      <td>Strain Name:A/seal/Mass/1/1980</td>
      <td>Protein Name:HA Hemagglutinin</td>
      <td>Gene Symbol:HA</td>
      <td>Segment:4</td>
      <td>Subtype:H7N7</td>
      <td>Host:Sea Mammal</td>
    </tr>
    <tr>
      <th>1</th>
      <td>gb:J02176|ncbiId:AAA43209.1|UniProtKB:P03454|O...</td>
      <td>gb:J02176</td>
      <td>ncbiId:AAA43209.1</td>
      <td>UniProtKB:P03454</td>
      <td>Organism:Influenza A virus (A/WSN/1933(H1N1))</td>
      <td>Strain Name:A/WSN/1933</td>
      <td>Protein Name:HA Hemagglutinin</td>
      <td>Gene Symbol:HA</td>
      <td>Segment:4</td>
      <td>Subtype:H1N1</td>
      <td>Host:Human</td>
    </tr>
    <tr>
      <th>2</th>
      <td>gb:CY021709|ncbiId:ABP49327.1|UniProtKB:A4U6V2...</td>
      <td>gb:CY021709</td>
      <td>ncbiId:ABP49327.1</td>
      <td>UniProtKB:A4U6V2</td>
      <td>Organism:Influenza A virus (A/AA/Huston/1945(H...</td>
      <td>Strain Name:A/AA/Huston/1945</td>
      <td>Protein Name:HA Hemagglutinin</td>
      <td>Gene Symbol:HA</td>
      <td>Segment:4</td>
      <td>Subtype:H1N1</td>
      <td>Host:Human</td>
    </tr>
    <tr>
      <th>3</th>
      <td>gb:CY020285|ncbiId:ABO38054.1|UniProtKB:A4GBW6...</td>
      <td>gb:CY020285</td>
      <td>ncbiId:ABO38054.1</td>
      <td>UniProtKB:A4GBW6</td>
      <td>Organism:Influenza A virus (A/AA/Marton/1943(H...</td>
      <td>Strain Name:A/AA/Marton/1943</td>
      <td>Protein Name:HA Hemagglutinin</td>
      <td>Gene Symbol:HA</td>
      <td>Segment:4</td>
      <td>Subtype:H1N1</td>
      <td>Host:Human</td>
    </tr>
    <tr>
      <th>4</th>
      <td>gb:CY083910|ncbiId:ADX98960.1|UniProtKB:F0TT51...</td>
      <td>gb:CY083910</td>
      <td>ncbiId:ADX98960.1</td>
      <td>UniProtKB:F0TT51</td>
      <td>Organism:Influenza A virus (A/Aalborg/INS132/2...</td>
      <td>Strain Name:A/Aalborg/INS132/2009</td>
      <td>Protein Name:HA Hemagglutinin</td>
      <td>Gene Symbol:HA</td>
      <td>Segment:4</td>
      <td>Subtype:H1N1</td>
      <td>Host:Human</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Remove prefix from each entry in the description based on spliting at the colon :</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">:</span>
    <span class="n">df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="n">c</span><span class="o">+</span><span class="s1">&#39;:&#39;</span><span class="p">)</span>

<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>description</th>
      <th>gb</th>
      <th>ncbiId</th>
      <th>UniProtKB</th>
      <th>Organism</th>
      <th>Strain Name</th>
      <th>Protein Name</th>
      <th>Gene Symbol</th>
      <th>Segment</th>
      <th>Subtype</th>
      <th>Host</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>gb:K00429|ncbiId:AAR96248.1|UniProtKB:Q6LEJ4|O...</td>
      <td>K00429</td>
      <td>AAR96248.1</td>
      <td>Q6LEJ4</td>
      <td>Influenza A virus (A/seal/Mass/1/1980(H7N7))</td>
      <td>A/seal/Mass/1/1980</td>
      <td>HA Hemagglutinin</td>
      <td>HA</td>
      <td>4</td>
      <td>H7N7</td>
      <td>Sea Mammal</td>
    </tr>
    <tr>
      <th>1</th>
      <td>gb:J02176|ncbiId:AAA43209.1|UniProtKB:P03454|O...</td>
      <td>J02176</td>
      <td>AAA43209.1</td>
      <td>03454</td>
      <td>Influenza A virus (A/WSN/1933(H1N1))</td>
      <td>A/WSN/1933</td>
      <td>HA Hemagglutinin</td>
      <td>HA</td>
      <td>4</td>
      <td>H1N1</td>
      <td>uman</td>
    </tr>
    <tr>
      <th>2</th>
      <td>gb:CY021709|ncbiId:ABP49327.1|UniProtKB:A4U6V2...</td>
      <td>CY021709</td>
      <td>ABP49327.1</td>
      <td>A4U6V2</td>
      <td>Influenza A virus (A/AA/Huston/1945(H1N1))</td>
      <td>A/AA/Huston/1945</td>
      <td>HA Hemagglutinin</td>
      <td>HA</td>
      <td>4</td>
      <td>H1N1</td>
      <td>uman</td>
    </tr>
    <tr>
      <th>3</th>
      <td>gb:CY020285|ncbiId:ABO38054.1|UniProtKB:A4GBW6...</td>
      <td>CY020285</td>
      <td>ABO38054.1</td>
      <td>A4GBW6</td>
      <td>Influenza A virus (A/AA/Marton/1943(H1N1))</td>
      <td>A/AA/Marton/1943</td>
      <td>HA Hemagglutinin</td>
      <td>HA</td>
      <td>4</td>
      <td>H1N1</td>
      <td>uman</td>
    </tr>
    <tr>
      <th>4</th>
      <td>gb:CY083910|ncbiId:ADX98960.1|UniProtKB:F0TT51...</td>
      <td>CY083910</td>
      <td>ADX98960.1</td>
      <td>F0TT51</td>
      <td>Influenza A virus (A/Aalborg/INS132/2009(H1N1))</td>
      <td>A/Aalborg/INS132/2009</td>
      <td>HA Hemagglutinin</td>
      <td>HA</td>
      <td>4</td>
      <td>H1N1</td>
      <td>uman</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</div>
<div class="section" id="process-sequences">
<h2>Process Sequences<a class="headerlink" href="#process-sequences" title="Permalink to this headline">¶</a></h2>
<p>Now we can return to the sequences and take a look at an example.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Seq(&#39;MNTQILVFIACVLIEAKGDKICLGHHAVANGTKVNTLTERGIEVVNATETVETA...ICI&#39;)
</pre></div>
</div>
</div>
</div>
<p>In order to begin efficiently working with these sequences, we need to convert them to an array and a numeric encoding.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">AAs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;Q&#39;</span><span class="p">,</span> <span class="s1">&#39;E&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">,</span>
        <span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;K&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;F&#39;</span><span class="p">,</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;W&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;V&#39;</span><span class="p">,</span> <span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;Z&#39;</span><span class="p">,</span> <span class="s1">&#39;J&#39;</span><span class="p">,</span> <span class="s1">&#39;U&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;Z&#39;</span>
    <span class="p">]</span>
<span class="n">vocabulary</span> <span class="o">=</span> <span class="p">{</span> <span class="n">aa</span><span class="p">:</span> <span class="n">idx</span> <span class="o">+</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">AAs</span><span class="p">))</span> <span class="p">}</span>

<span class="n">seqs_pre</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">:</span>
    <span class="n">seqs_pre</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="p">[</span><span class="n">vocabulary</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">s</span><span class="p">]</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The model is ultimately trained by asking it to predict the identiy of an amino acid at position <span class="math notranslate nohighlight">\(i\)</span> based on the sequence before <span class="math notranslate nohighlight">\(X[:i]\)</span> or after <span class="math notranslate nohighlight">\(X[i+1:]\)</span> it. We need a utility function to take each sequence in our dataset and split it at each possible position in order to create the training dataset. For this purpose, we are only going to train on a small fraction of the dataset so that we can keep the memory usage manageable. We will use numpy to generate a list of random indicies that we will use to randomly select a subset of our data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">split_and_pad</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span><span class="n">seq_len</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;For each seq, split at each possible amino acid position </span>
<span class="sd">    into pre and post sequences&#39;&#39;&#39;</span>

    <span class="c1"># Split each sequence at each possible index i</span>
    <span class="n">X_pre</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="p">[:</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">seqs</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">))]</span>
    <span class="n">X_post</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">seqs</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">))]</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">s</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">seqs</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">))])</span>

    <span class="c1"># Pad X_pre and X_post with pre or post padding respectively</span>
    <span class="n">X_pre</span> <span class="o">=</span> <span class="n">tf</span><span class="o">.</span><span class="n">keras</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">pad_sequences</span><span class="p">(</span>
        <span class="n">X_pre</span><span class="p">,</span> <span class="n">maxlen</span><span class="o">=</span><span class="n">seq_len</span><span class="p">,</span> <span class="n">padding</span><span class="o">=</span><span class="s1">&#39;pre&#39;</span><span class="p">,</span> <span class="n">truncating</span><span class="o">=</span><span class="s1">&#39;pre&#39;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="mi">0</span>
    <span class="p">)</span>
    <span class="n">X_post</span> <span class="o">=</span> <span class="n">tf</span><span class="o">.</span><span class="n">keras</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">pad_sequences</span><span class="p">(</span>
        <span class="n">X_post</span><span class="p">,</span> <span class="n">maxlen</span><span class="o">=</span><span class="n">seq_len</span><span class="p">,</span> <span class="n">padding</span><span class="o">=</span><span class="s1">&#39;post&#39;</span><span class="p">,</span> <span class="n">truncating</span><span class="o">=</span><span class="s1">&#39;post&#39;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="mi">0</span>
    <span class="p">)</span>

    <span class="c1"># Flip X_post values</span>
    <span class="n">X_post</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">X_post</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Stack X into single array</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">stack</span><span class="p">([</span><span class="n">X_pre</span><span class="p">,</span> <span class="n">X_post</span><span class="p">],</span><span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">X</span><span class="p">,</span> <span class="n">y</span>

<span class="c1"># Generate random selection</span>
<span class="n">selector</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seqs_pre</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="n">seqs_pre</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">selector</span><span class="p">]</span>

<span class="n">X</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">split_and_pad</span><span class="p">(</span><span class="n">subset</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">length</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">y</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(565017, 575, 2) (565017,)
</pre></div>
</div>
</div>
</div>
<p>We can now pass our data into a tf.Dataset object to facilitate training.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dataset</span> <span class="o">=</span> <span class="n">tf</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">Dataset</span><span class="o">.</span><span class="n">from_tensor_slices</span><span class="p">(((</span><span class="n">X</span><span class="p">[</span><span class="o">...</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span><span class="n">X</span><span class="p">[</span><span class="o">...</span><span class="p">,</span><span class="mi">1</span><span class="p">]),</span><span class="n">y</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>If we look at a single <code class="docutils literal notranslate"><span class="pre">take</span></code> from the dataset, we can see a padded sequence with numeric encoding.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">dataset</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="mi">1</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>((&lt;tf.Tensor: shape=(575,), dtype=int32, numpy=
array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0], dtype=int32)&gt;, &lt;tf.Tensor: shape=(575,), dtype=int32, numpy=
array([ 0,  0,  0,  0,  0,  0,  0,  0,  0,  9,  3,  9, 17,  3, 16, 12, 18,
        7, 14, 18,  3, 13, 22, 12, 18, 12,  7,  1, 13, 13,  9,  1, 12, 21,
       12, 18, 18,  1, 21, 19, 18, 24,  9, 18, 12,  9, 16, 24,  9,  7,  9,
       18,  5, 12, 11, 21,  7, 18,  9,  5,  5, 17, 11, 12, 17,  1,  5,  5,
       18, 24, 16, 15, 24,  4, 24, 19,  7, 14, 17, 21, 18,  5, 13,  3,  5,
       14, 14,  3, 11,  8, 24,  6,  5,  6,  3,  7, 14,  7, 12,  5, 11,  1,
       14,  4, 11, 12, 16, 12, 17, 21, 11,  4, 24, 12, 14, 17, 21, 14, 18,
        4,  8,  6,  4, 12, 19, 17,  5, 14,  5, 13, 12, 21, 12, 12,  5,  1,
       14, 24, 19, 22, 21,  4, 12,  6,  7,  4,  5, 13, 11, 11, 14, 12, 14,
        5,  9, 17, 17,  5, 12, 14, 14,  6,  5, 17,  7, 21,  1,  5,  6, 16,
       19, 14, 13, 11,  4,  9,  9, 18, 14, 21, 11, 14, 19, 21,  7,  4,  9,
        1, 11, 16, 19, 18,  5, 11,  4,  1,  1, 24,  7, 18,  7, 16,  5, 14,
       18,  8,  8, 24,  7,  6, 22,  7,  4, 21, 13,  7, 16, 22,  7,  7,  5,
        9,  6,  7,  1,  9,  1,  7,  6, 12,  7, 17, 11, 17, 17, 17,  5, 17,
       16, 15, 18, 14, 17, 12,  7, 19,  1, 12, 21, 12, 11, 14, 18, 11, 21,
       24, 11, 15,  3,  5,  7,  9, 19, 12, 15,  8,  9, 14,  8,  6, 15, 13,
       18, 18, 14,  9,  1,  7,  9, 15, 19, 16,  3, 17, 19, 14,  3, 14,  7,
       24,  5, 21,  5, 18, 11, 13,  9, 19, 18,  4,  7, 11, 11, 21,  9, 11,
       24,  1, 24,  5, 15,  1,  9,  6, 14,  7, 14, 18,  5,  6,  8,  9,  1,
        4, 14, 15, 11, 12,  9, 19, 22,  6,  6,  4,  9, 17,  7, 18, 16,  7,
       14,  9, 11, 18, 17, 19,  1,  9, 11, 15, 21, 12, 17, 16, 14, 12, 19,
       18, 19,  7,  9, 18,  9, 24, 19, 19, 15, 14, 16, 24, 12, 11, 19, 16,
        5,  1,  5,  4, 14, 15,  8,  8,  9,  7, 22, 12, 21, 12, 12,  4,  5,
       16, 14, 19, 14, 14, 24,  7, 11, 11,  9, 19, 15, 24,  1, 14,  4, 11,
       11,  9, 12, 22, 21, 21, 14, 17,  6,  6, 18, 18, 14,  7, 16, 24, 15,
        3,  1,  1, 18, 21,  7, 12, 18,  1,  5,  8,  4, 18, 22, 18,  4, 11,
       15,  9,  9, 16,  9, 11,  5,  6,  8, 14,  9, 17, 18, 12, 12,  8, 11,
       12,  5,  5, 24,  4, 14,  6, 14,  7, 15, 24,  3, 12,  4, 14,  1, 15,
       14,  1, 11,  5, 21,  9, 24, 18, 22,  5, 15, 21, 14,  9,  6,  5,  4,
        3, 13, 15, 14,  7, 12, 12, 22,  7,  1, 21, 18,  3,  4, 11, 12,  9,
       12, 15, 11, 21,  7, 14, 12,  4,  3, 12, 11,  7, 14,  8, 19, 11,  5,
       12,  9,  4, 16,  1,  8, 19, 21, 19, 21, 14, 11,  5, 13,  9, 19,  4,
       21, 16,  5, 19, 18, 14, 14,  1,  8, 24,  7,  9,  3,  9,  8,  4, 18,
       11, 21, 12, 18,  9, 19,  1,  6, 12, 12, 21,  9, 11,  5],
      dtype=int32)&gt;), &lt;tf.Tensor: shape=(), dtype=int64, numpy=13&gt;)
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="build-model">
<h2>Build Model<a class="headerlink" href="#build-model" title="Permalink to this headline">¶</a></h2>
<p>In the paper, they test a several model architectures and a variety of hyperparameters before ultimately selecting a BiLSTM architecture with two hidden layers of 512 units each. The model receives sequence context for before and after the missing amino acid.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">tensorflow.keras</span> <span class="kn">import</span> <span class="n">Input</span>
<span class="kn">from</span> <span class="nn">tensorflow.keras</span> <span class="kn">import</span> <span class="n">layers</span>
<span class="kn">from</span> <span class="nn">tensorflow.keras.models</span> <span class="kn">import</span> <span class="n">Model</span>

<span class="k">def</span> <span class="nf">BiLSTM</span><span class="p">(</span><span class="n">seq_len</span><span class="p">,</span> 
           <span class="n">vocab_size</span><span class="p">,</span>
           <span class="n">embedding_dim</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
           <span class="n">hidden_dim</span><span class="o">=</span><span class="mi">256</span><span class="p">,</span>
           <span class="n">n_hidden</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>

    <span class="c1"># Define input layers</span>
    <span class="n">input_pre</span> <span class="o">=</span> <span class="n">Input</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">seq_len</span><span class="p">,))</span>
    <span class="n">input_post</span> <span class="o">=</span> <span class="n">Input</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">seq_len</span><span class="p">,))</span>

    <span class="c1"># Generate embedding layer and run both inputs through</span>
    <span class="n">embed</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">Embedding</span><span class="p">(</span><span class="n">vocab_size</span><span class="p">,</span> <span class="n">embedding_dim</span><span class="p">,</span> <span class="n">input_length</span><span class="o">=</span><span class="n">seq_len</span><span class="p">)</span>
    <span class="n">xpre</span> <span class="o">=</span> <span class="n">embed</span><span class="p">(</span><span class="n">input_pre</span><span class="p">)</span>
    <span class="n">xpost</span> <span class="o">=</span> <span class="n">embed</span><span class="p">(</span><span class="n">input_post</span><span class="p">)</span>

    <span class="c1"># Generate an lstm layer for each hidden dimension</span>
    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_hidden</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">lstm</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">LSTM</span><span class="p">(</span><span class="n">hidden_dim</span><span class="p">,</span> <span class="n">return_sequences</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">xpre</span> <span class="o">=</span> <span class="n">lstm</span><span class="p">(</span><span class="n">xpre</span><span class="p">)</span>
        <span class="n">xpost</span> <span class="o">=</span> <span class="n">lstm</span><span class="p">(</span><span class="n">xpost</span><span class="p">)</span>

    <span class="c1"># Define final LSTM layer</span>
    <span class="n">lstm</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">LSTM</span><span class="p">(</span><span class="n">hidden_dim</span><span class="p">)</span>
    <span class="n">xpre</span> <span class="o">=</span> <span class="n">lstm</span><span class="p">(</span><span class="n">xpre</span><span class="p">)</span>
    <span class="n">xpost</span> <span class="o">=</span> <span class="n">lstm</span><span class="p">(</span><span class="n">xpost</span><span class="p">)</span>

    <span class="c1"># Concatenate pre and post before sending to dense</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">xpre</span><span class="p">,</span> <span class="n">xpost</span><span class="p">],</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;embed_layer&#39;</span><span class="p">)</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">Dense</span><span class="p">(</span><span class="n">vocab_size</span><span class="p">)(</span><span class="n">x</span><span class="p">)</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">Activation</span><span class="p">(</span><span class="s1">&#39;softmax&#39;</span><span class="p">)(</span><span class="n">x</span><span class="p">)</span>

    <span class="n">bilstm</span> <span class="o">=</span> <span class="n">Model</span><span class="p">(</span><span class="n">inputs</span><span class="o">=</span><span class="p">[</span><span class="n">input_pre</span><span class="p">,</span> <span class="n">input_post</span><span class="p">],</span> <span class="n">outputs</span><span class="o">=</span><span class="n">output</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">bilstm</span>

<span class="n">model</span> <span class="o">=</span> <span class="n">BiLSTM</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="nb">len</span><span class="p">(</span><span class="n">vocabulary</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Model: &quot;model&quot;
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_1 (InputLayer)            [(None, 575)]        0                                            
__________________________________________________________________________________________________
input_2 (InputLayer)            [(None, 575)]        0                                            
__________________________________________________________________________________________________
embedding (Embedding)           (None, 575, 20)      500         input_1[0][0]                    
                                                                 input_2[0][0]                    
__________________________________________________________________________________________________
lstm (LSTM)                     (None, 575, 256)     283648      embedding[0][0]                  
                                                                 embedding[1][0]                  
__________________________________________________________________________________________________
lstm_1 (LSTM)                   (None, 256)          525312      lstm[0][0]                       
                                                                 lstm[1][0]                       
__________________________________________________________________________________________________
embed_layer (Concatenate)       (None, 512)          0           lstm_1[0][0]                     
                                                                 lstm_1[1][0]                     
__________________________________________________________________________________________________
dense (Dense)                   (None, 25)           12825       embed_layer[0][0]                
__________________________________________________________________________________________________
activation (Activation)         (None, 25)           0           dense[0][0]                      
==================================================================================================
Total params: 822,285
Trainable params: 822,285
Non-trainable params: 0
__________________________________________________________________________________________________
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">tensorflow.keras.optimizers</span> <span class="kn">import</span> <span class="n">Adam</span>

<span class="n">n_epochs</span> <span class="o">=</span> <span class="mi">6</span>
<span class="n">batch_size</span> <span class="o">=</span> <span class="mi">100</span>

<span class="n">optimizer</span> <span class="o">=</span> <span class="n">Adam</span><span class="p">(</span><span class="n">learning_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">beta_1</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">beta_2</span><span class="o">=</span><span class="mf">0.999</span><span class="p">,</span> <span class="n">amsgrad</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">loss</span><span class="o">=</span><span class="s1">&#39;sparse_categorical_crossentropy&#39;</span><span class="p">,</span>
              <span class="n">optimizer</span><span class="o">=</span><span class="n">optimizer</span><span class="p">,</span>
              <span class="n">metrics</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;accuracy&#39;</span><span class="p">])</span>

<span class="n">MODEL_PATH</span> <span class="o">=</span> <span class="s1">&#39;flu-sequence-20210330-epoch</span><span class="si">{epoch:02d}</span><span class="s1">.h5&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="mi">2000</span><span class="p">)</span><span class="o">.</span><span class="n">batch</span><span class="p">(</span><span class="n">batch_size</span><span class="p">),</span>
          <span class="n">epochs</span><span class="o">=</span><span class="n">n_epochs</span><span class="p">,</span>
          <span class="n">callbacks</span><span class="o">=</span><span class="p">[</span><span class="n">tf</span><span class="o">.</span><span class="n">keras</span><span class="o">.</span><span class="n">callbacks</span><span class="o">.</span><span class="n">ModelCheckpoint</span><span class="p">(</span>
            <span class="n">MODEL_PATH</span><span class="p">,</span>
            <span class="n">monitor</span><span class="o">=</span><span class="s1">&#39;accuracy&#39;</span><span class="p">,</span>
            <span class="n">verbose</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">save_best_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 1/6
1307/5652 [=====&gt;........................] - ETA: 11:20:56 - loss: 2.7632 - accuracy: 0.1359
</pre></div>
</div>
</div>
</div>
<p>The model was ultimately trained for 6 epochs to get started on the training process.</p>
</div>
<div class="section" id="sample-embeddings">
<h2>Sample Embeddings<a class="headerlink" href="#sample-embeddings" title="Permalink to this headline">¶</a></h2>
<p>We are now ready to use hidden layer to extract the semantic embeddings of sequences.</p>
<div class="cell tag_hide-cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Download weights file from bucket</span>
<span class="o">!</span>wget https://storage.googleapis.com/datasets-spring2021/flu-sequence-20210330-epoch2-04.h5
</pre></div>
</div>
</div>
</div>
<div class="cell tag_hide-cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">load_weights</span><span class="p">(</span><span class="s1">&#39;flu-sequence-20210330-epoch2-04.h5&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Retrieve hidden layer to use for the embedding</span>
<span class="n">hidden</span> <span class="o">=</span> <span class="n">tf</span><span class="o">.</span><span class="n">keras</span><span class="o">.</span><span class="n">backend</span><span class="o">.</span><span class="n">function</span><span class="p">(</span>
    <span class="n">inputs</span><span class="o">=</span><span class="n">model</span><span class="o">.</span><span class="n">input</span><span class="p">,</span>
    <span class="n">outputs</span><span class="o">=</span><span class="n">model</span><span class="o">.</span><span class="n">get_layer</span><span class="p">(</span><span class="s1">&#39;embed_layer&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">output</span>
<span class="p">)</span>

<span class="c1"># Run predictions in batches to manage memory</span>
<span class="n">x_embed_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="mi">5000</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">5000</span><span class="p">,</span> <span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">x_embed_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hidden</span><span class="p">([</span><span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">end</span><span class="p">,:,</span><span class="mi">0</span><span class="p">],</span><span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">end</span><span class="p">,:,</span><span class="mi">0</span><span class="p">]]))</span>
<span class="n">x_embed_cat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">x_embed_list</span><span class="p">)</span>
<span class="n">x_embed_cat</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>..................................................................................................................
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(565017, 512)
</pre></div>
</div>
</div>
</div>
<p>Currently the first axis is representative of all sequences concatenated together. We can reshape to introduce the sample axis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">x_embed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">stack</span><span class="p">([</span>
    <span class="n">x_embed_cat</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">575</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,(</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">//</span><span class="mi">575</span><span class="p">)</span><span class="o">*</span><span class="mi">575</span><span class="p">,</span><span class="mi">575</span><span class="p">)</span>
<span class="p">])</span>
<span class="n">x_embed</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(982, 575, 512)
</pre></div>
</div>
</div>
</div>
<p>Finally, we take find the average value of the sequence for each of the 512 dimensions in the last axis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xmean</span> <span class="o">=</span> <span class="n">x_embed</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="n">xmean</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(982, 512)
</pre></div>
</div>
</div>
</div>
<div class="cell tag_hide-cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">savez_compressed</span><span class="p">(</span><span class="s1">&#39;xmean.npz&#39;</span><span class="p">,</span><span class="n">xmean</span><span class="o">=</span><span class="n">xmean</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The paper elects to use the UMAP algorithm to reduce the data from 512 dimensions to 2 which can be easily visualized. We will use the <a class="reference external" href="https://github.com/lmcinnes/umap">python package</a> that was published with the original paper.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reducer</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">()</span>
<span class="n">embedding</span> <span class="o">=</span> <span class="n">reducer</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">xmean</span><span class="p">)</span>
<span class="n">embedding</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(982, 2)
</pre></div>
</div>
</div>
</div>
<p>In order to produce visualizations that we can begin to interpret, we will link our embeddings back to the original metadata that we collected about each sequence.</p>
<div class="cell tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_embed</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">982</span><span class="p">]</span>
<span class="n">df_embed</span><span class="p">[</span><span class="s1">&#39;embed_0&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">embedding</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span>
<span class="n">df_embed</span><span class="p">[</span><span class="s1">&#39;embed_1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">embedding</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>For an initial visualization, we will plot each sequence and color code according to the viral subtype.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">cat</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">df_embed</span><span class="p">[</span><span class="s1">&#39;Subtype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()):</span>
    <span class="n">indx</span> <span class="o">=</span> <span class="n">df_embed</span><span class="p">[</span><span class="s1">&#39;Subtype&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cat</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">df_embed</span><span class="p">[</span><span class="s1">&#39;embed_0&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">indx</span><span class="p">],</span><span class="n">df_embed</span><span class="p">[</span><span class="s1">&#39;embed_1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">indx</span><span class="p">],</span><span class="n">label</span><span class="o">=</span><span class="n">cat</span><span class="p">)</span>
    
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">borderaxespad</span><span class="o">=</span><span class="mf">0.</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;matplotlib.legend.Legend at 0x2ab7b3da8e20&gt;
</pre></div>
</div>
<img alt="../_images/sequence-model_45_1.png" src="../_images/sequence-model_45_1.png" />
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "conda-env-bebi205-py"
        },
        kernelOptions: {
            kernelName: "conda-env-bebi205-py",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'conda-env-bebi205-py'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="segmentation.html" title="previous page">Cell Segmentation</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By David Van Valen and Morgan Schwartz<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>